AuthorizeServerConn failed certificate validation

Hi team,

I running a consul server cluster of 3 nodes VM.
My consul version was 1.9.8, and I deciced to upgrade my cluster to 1.10.3 version and in here is what I saw in the logs:

```
from=10.0.104.13:57854 operation="raft RPC"
2021-10-02T13:53:24.695Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-02T13:53:24.696Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.0.104.12:8300 [Candidate]" term=168
2021-10-02T13:53:24.701Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-02T13:53:24.701Z [DEBUG] agent.server.raft: vote granted: from=2afb25d5-8234-e9fc-dcef-babf75741651 term=168 tally=1
2021-10-02T13:53:24.707Z [ERROR] agent.server.raft: failed to make requestVote RPC: target="{Voter 24a6c35e-9bab-7bae-9a31-ae94400ac1ef 10.0.104.13:8300}" error=EOF
2021-10-02T13:53:24.708Z [ERROR] agent.server.raft: failed to make requestVote RPC: target="{Voter 3840bd0b-0980-8c1e-47c7-73bfef3bb5b9 10.0.104.10:8300}" error=EOF
2021-10-02T13:53:24.803Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.euw1.consul: %!w(<nil>): from=10.0.104.10:48992 operation="raft RPC"
2021-10-02T13:53:24.913Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.euw1.consul: %!w(<nil>): from=10.0.104.13:57858 operation="raft RPC"
2021-10-02T13:53:26.045Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.euw1.consul: %!w(<nil>): from=10.0.104.13:57862 operation="raft RPC"
2021-10-02T13:53:26.083Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.euw1.consul: %!w(<nil>): from=10.0.104.10:48998 operation="raft RPC"
```

Before making the upgrade I shut them all down and backup the data just in case as it's not a production server.
After googling and not able to make things works, I revert my install with the old backup.
I decided then to upgrade my 1.9.8 to 1.9.10 and unfortunately I got the same errors!

#### Reproduction Steps

Steps to reproduce this issue, eg:

My consul config is:

```
{
    "acl": {
        "default_policy": "deny",
        "down_policy": "extend-cache",
        "enable_token_persistence": true,
        "enabled": true,
        "tokens": {
            "agent": "xxxx",
            "default": "xxxx",
            "master": "xxxx"
        }
    },
    "advertise_addr": "x.x.x.x",
    "auto_encrypt": {
        "allow_tls": true
    },
    "bind_addr": "0.0.0.0",
    "bootstrap_expect": 3,
    "ca_file": "/etc/consul.d/.ssl/euw1-consul-agent-ca.pem",
    "cert_file": "/etc/consul.d/.ssl/euw1-server-consul.pem",
    "client_addr": "0.0.0.0",
    "connect": {
        "enabled": true
    },
    "data_dir": "/data/consul",
    "datacenter": "euw1",
    "disable_update_check": true,
    "enable_central_service_config": true,
    "encrypt": "9a896L5GCtRFfLkyyUfLsxOjIbX4X/2b6rx9ipz74c0=",
    "key_file": "/etc/consul.d/.ssl/euw1-server-consul-key.pem",
    "log_level": "DEBUG",
    "node_name": "x.x.x.x",
    "performance": {
        "raft_multiplier": 1
    },
    "ports": {
        "grpc": 8502,
        "https": 8501
    },
    "retry_join": [
        "x.x.x.x",
        "x.x.x.x",
        "x.x.x.x"
    ],
    "server": true,
    "telemetry": {
        "prometheus_retention_time": "30s"
    },
    "ui_config": {
        "enabled": true,
        "metrics_provider": "prometheus"
    },
    "verify_incoming": false,
    "verify_outgoing": true,
    "verify_server_hostname": true
}
```

### Consul info Server

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease = 
	revision = 34a6478b
	version = 1.9.10
consul:
	acl = legacy
	bootstrap = false
	known_datacenters = 1
	leader = false
	leader_addr = 
	server = true
raft:
	applied_index = 32774
	commit_index = 0
	fsm_pending = 0
	last_contact = never
	last_log_index = 45589
	last_log_term = 251
	last_snapshot_index = 32774
	last_snapshot_term = 116
	latest_configuration = [{Suffrage:Voter ID:2afb25d5-8234-e9fc-dcef-babf75741651 Address:10.0.104.12:8300} {Suffrage:Voter ID:24a6c35e-9bab-7bae-9a31-ae94400ac1ef Address:10.0.104.13:8300} {Suffrage:Voter ID:3840bd0b-0980-8c1e-47c7-73bfef3bb5b9 Address:10.0.104.10:8300}]
	latest_configuration_index = 0
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Candidate
	term = 993
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 128
	max_procs = 1
	os = linux
	version = go1.15.15
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 34
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 3530
	members = 10
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 47
	members = 3
	query_queue = 0
	query_time = 1
```

</details>

### Logs fragments

```
==> Starting Consul agent...
           Version: '1.9.10'
           Node ID: '2afb25d5-8234-e9fc-dcef-babf75741651'
         Node name: 'x.x.x.x'
        Datacenter: 'euw1' (Segment: '<all>')
            Server: true (Bootstrap: false)
       Client Addr: [0.0.0.0] (HTTP: 8500, HTTPS: 8501, gRPC: 8502, DNS: 8600)
      Cluster Addr: 10.0.104.12 (LAN: 8301, WAN: 8302)
           Encrypt: Gossip: true, TLS-Outgoing: true, TLS-Incoming: false, Auto-Encrypt-TLS: true
```

### Operating system and Environment details

Debian 10

Can someone tell me what's going on as the cluster was down after both upgrades?
I don't have any other meaningful logs to provide but feel free to ask more if necessary.
The firewal ports opened between the 3 servers are 8300-8302 (tcp), 8501-8502 (tcp), 8301-8302 (udp).

Ran it directly on the upgraded server, it's trying to start its own election:
```
2021-10-04T18:30:11.243Z [DEBUG] agent.http: Request finished: method=GET url=/v1/agent/services from=127.0.0.1:51366 latency=162.457µs
2021-10-04T18:30:11.243Z [DEBUG] agent.http: Request finished: method=GET url=/v1/agent/checks from=127.0.0.1:51366 latency=115.428µs
2021-10-04T18:30:13.020Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:30:14.070Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:30:14.670Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:30:14.670Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1845
2021-10-04T18:30:14.676Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:30:14.676Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1845 tally=1
2021-10-04T18:30:16.926Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
2021-10-04T18:30:17.476Z [WARN]  agent.server.rpc: a TLS certificate with a CommonName of server.playground.plgconsul is required for this operation: from=10.201.20.22:45488 operation="raft RPC"
2021-10-04T18:30:17.788Z [WARN]  agent.server.rpc: a TLS certificate with a CommonName of server.playground.plgconsul is required for this operation: from=10.201.20.22:45490 operation="raft RPC"
2021-10-04T18:30:20.092Z [DEBUG] agent.server: Cannot upgrade to new ACLs: leaderMode=3 mode=2 found=true leader=
2021-10-04T18:30:21.754Z [DEBUG] agent.server.memberlist.lan: memberlist: Stream connection from=10.201.20.21:46676
2021-10-04T18:30:23.026Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:30:24.070Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:30:24.524Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:30:24.524Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1846
2021-10-04T18:30:24.530Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:30:24.530Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1846 tally=1
2021-10-04T18:30:25.055Z [DEBUG] agent.server.memberlist.lan: memberlist: Stream connection from=10.201.20.45:47134
2021-10-04T18:30:25.861Z [ERROR] agent: Coordinate update error: error="No cluster leader"
2021-10-04T18:30:25.929Z [DEBUG] agent.server.memberlist.lan: memberlist: Stream connection from=10.201.20.44:38484
2021-10-04T18:30:26.926Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
2021-10-04T18:30:28.034Z [WARN]  agent.server.rpc: a TLS certificate with a CommonName of server.playground.plgconsul is required for this operation: from=10.201.20.22:45498 operation="raft RPC"
2021-10-04T18:30:28.278Z [WARN]  agent.server.rpc: a TLS certificate with a CommonName of server.playground.plgconsul is required for this operation: from=10.201.20.22:45500 operation="raft RPC"
2021-10-04T18:30:32.127Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:30:32.127Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1847
2021-10-04T18:30:32.134Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:30:32.134Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1847 tally=1
2021-10-04T18:30:33.032Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:30:34.071Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:30:34.644Z [DEBUG] agent.server.memberlist.wan: memberlist: Initiating push/pull sync with: nmdserver-plg-02.playground 10.201.20.21:8302
2021-10-04T18:30:34.972Z [DEBUG] agent.server.memberlist.lan: memberlist: Initiating push/pull sync with: nmdclient-naeast-a-plg-t5dn 10.201.20.45:8301
2021-10-04T18:30:36.927Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
2021-10-04T18:30:37.240Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:30:37.240Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1848
2021-10-04T18:30:37.247Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:30:37.247Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1848 tally=1
2021-10-04T18:30:38.280Z [WARN]  agent.server.rpc: a TLS certificate with a CommonName of server.playground.plgconsul is required for this operation: from=10.201.20.22:45518 operation="raft RPC"
2021-10-04T18:30:39.315Z [WARN]  agent.server.rpc: a TLS certificate with a CommonName of server.playground.plgconsul is required for this operation: from=10.201.20.22:45520 operation="raft RPC"
2021-10-04T18:30:40.756Z [ERROR] agent.anti_entropy: failed to sync remote state: error="No cluster leader"
2021-10-04T18:30:41.244Z [DEBUG] agent.http: Request finished: method=GET url=/v1/agent/services from=127.0.0.1:51366 latency=153.128µs
2021-10-04T18:30:41.245Z [DEBUG] agent.http: Request finished: method=GET url=/v1/agent/checks from=127.0.0.1:51366 latency=114.584µs
2021-10-04T18:30:43.038Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:30:43.088Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:30:43.088Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1849
2021-10-04T18:30:43.102Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:30:43.102Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1849 tally=1
```

Logs in the case of 1.9.10:
```
2021-10-04T18:42:52.291Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.playground.plgconsul: %!w(<nil>): from=10.201.20.22:46596 operation="raft RPC"
2021-10-04T18:42:52.743Z [DEBUG] agent.server: Cannot upgrade to new ACLs: leaderMode=3 mode=2 found=true leader=
2021-10-04T18:42:53.258Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:42:54.219Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
2021-10-04T18:42:55.121Z [DEBUG] agent.server.memberlist.lan: memberlist: Stream connection from=10.201.20.45:48014
2021-10-04T18:42:55.142Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:42:56.432Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.playground.plgconsul: %!w(<nil>): from=10.201.20.22:46600 operation="raft RPC"
2021-10-04T18:42:56.594Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:42:56.594Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1944
2021-10-04T18:42:56.601Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:42:56.601Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1944 tally=1
2021-10-04T18:42:59.401Z [DEBUG] agent.server.memberlist.lan: memberlist: Stream connection from=10.201.20.22:55750
2021-10-04T18:43:02.827Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:43:02.827Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1945
2021-10-04T18:43:02.834Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:43:02.834Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1945 tally=1
2021-10-04T18:43:03.258Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.playground.plgconsul: %!w(<nil>): from=10.201.20.22:46612 operation="raft RPC"
2021-10-04T18:43:03.258Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:43:04.219Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
2021-10-04T18:43:05.147Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:43:05.543Z [DEBUG] agent.server: Cannot upgrade to new ACLs: leaderMode=3 mode=2 found=true leader=
2021-10-04T18:43:06.678Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.playground.plgconsul: %!w(<nil>): from=10.201.20.22:46622 operation="raft RPC"
2021-10-04T18:43:08.191Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:43:08.191Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1946
2021-10-04T18:43:08.197Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:43:08.197Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1946 tally=1
2021-10-04T18:43:11.315Z [DEBUG] agent.http: Request finished: method=GET url=/v1/agent/services from=127.0.0.1:52288 latency=154.638µs
2021-10-04T18:43:11.316Z [DEBUG] agent.http: Request finished: method=GET url=/v1/agent/checks from=127.0.0.1:52288 latency=141.052µs
2021-10-04T18:43:12.273Z [ERROR] agent: Coordinate update error: error="No cluster leader"
2021-10-04T18:43:13.259Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:43:14.220Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
2021-10-04T18:43:14.371Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.playground.plgconsul: %!w(<nil>): from=10.201.20.22:46630 operation="raft RPC"
2021-10-04T18:43:15.153Z [DEBUG] agent: Check status updated: check=_nomad-check-73f7361be4d3ffd00dbe0214ebcc33b279f9baa4 status=passing
2021-10-04T18:43:15.478Z [ERROR] agent.anti_entropy: failed to sync remote state: error="No cluster leader"
2021-10-04T18:43:16.923Z [WARN]  agent.server.rpc: AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of server.playground.plgconsul: %!w(<nil>): from=10.201.20.22:46640 operation="raft RPC"
2021-10-04T18:43:17.660Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:43:17.660Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1947
2021-10-04T18:43:17.666Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:43:17.666Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1947 tally=1
2021-10-04T18:43:23.213Z [WARN]  agent.server.raft: Election timeout reached, restarting election
2021-10-04T18:43:23.213Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.201.20.20:8300 [Candidate]" term=1948
2021-10-04T18:43:23.222Z [DEBUG] agent.server.raft: votes: needed=2
2021-10-04T18:43:23.222Z [DEBUG] agent.server.raft: vote granted: from=f2175ca6-74c0-7b0a-01ca-e1af1e89a1c0 term=1948 tally=1
2021-10-04T18:43:23.259Z [DEBUG] agent: Check status updated: check=_nomad-check-b601bb092df00ce526aed5d4ec66fc0885116be4 status=passing
2021-10-04T18:43:24.220Z [DEBUG] agent: Check status updated: check=_nomad-check-ed687cd3e1411e1db0e2a4152cae62a8baa68b80 status=passing
```

--- a/.changelog/11255.txt
+++ b/.changelog/11255.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+rpc: only attempt to authorize the DNSName in the client cert when verify_incoming_rpc=true
+```

--- a/tlsutil/config.go
+++ b/tlsutil/config.go
@@ -904,10 +904,10 @@ type TLSConn interface {
 // presented is signed by the Agent TLS CA, and has a DNSName that matches the
 // local ServerSNI name.
 //
-// Note this check is only performed if VerifyServerHostname is enabled, otherwise
-// it does no authorization.
+// Note this check is only performed if VerifyServerHostname and VerifyIncomingRPC
+// are both enabled, otherwise it does no authorization.
 func (c *Configurator) AuthorizeServerConn(dc string, conn TLSConn) error {
-	if !c.VerifyServerHostname() {
+	if !c.VerifyIncomingRPC() || !c.VerifyServerHostname() {
 		return nil
 	}
 
@@ -936,6 +936,9 @@ func (c *Configurator) AuthorizeServerConn(dc string, conn TLSConn) error {
 			return nil
 		}
 		errs = multierror.Append(errs, err)
+	}
+	if errs == nil {
+		errs = fmt.Errorf("no verified chains")
 	}
 	return fmt.Errorf("AuthorizeServerConn failed certificate validation for certificate with a SAN.DNSName of %v: %w", expected, errs)
 

--- a/tlsutil/config_test.go
+++ b/tlsutil/config_test.go
@@ -1271,7 +1271,7 @@ func TestConfigurator_AutoEncryptCert(t *testing.T) {
 	require.Equal(t, int64(4679716209), c.AutoEncryptCert().NotAfter.Unix())
 }
 
-func TestConfigurator_AuthorizeServerConn_Error(t *testing.T) {
+func TestConfigurator_AuthorizeServerConn(t *testing.T) {
 	caPEM, caPK, err := GenerateCA(CAOpts{Days: 5, Domain: "consul"})
 	require.NoError(t, err)
 
@@ -1280,10 +1280,28 @@ func TestConfigurator_AuthorizeServerConn_Error(t *testing.T) {
 	err = ioutil.WriteFile(caPath, []byte(caPEM), 0600)
 	require.NoError(t, err)
 
+	// Cert and key are not used, but required to get past validateConfig
+	signer, err := ParseSigner(caPK)
+	require.NoError(t, err)
+	pub, pk, err := GenerateCert(CertOpts{
+		Signer: signer,
+		CA:     caPEM,
+	})
+	require.NoError(t, err)
+	certFile := filepath.Join("cert.pem")
+	err = ioutil.WriteFile(certFile, []byte(pub), 0600)
+	require.NoError(t, err)
+	keyFile := filepath.Join("cert.key")
+	err = ioutil.WriteFile(keyFile, []byte(pk), 0600)
+	require.NoError(t, err)
+
 	cfg := Config{
 		VerifyServerHostname: true,
+		VerifyIncomingRPC:    true,
 		Domain:               "consul",
 		CAFile:               caPath,
+		CertFile:             certFile,
+		KeyFile:              keyFile,
 	}
 	c, err := NewConfigurator(cfg, hclog.New(nil))
 	require.NoError(t, err)
@@ -1367,6 +1385,22 @@ func TestConfigurator_AuthorizeServerConn_Error(t *testing.T) {
 		err = c.AuthorizeServerConn("dc1", s)
 		testutil.RequireErrorContains(t, err, "certificate specifies an incompatible key usage")
 	})
+
+	t.Run("disabled by verify_incoming_rpc", func(t *testing.T) {
+		cfg := Config{
+			VerifyServerHostname: true,
+			VerifyIncomingRPC:    false,
+			Domain:               "consul",
+			CAFile:               caPath,
+		}
+		c, err := NewConfigurator(cfg, hclog.New(nil))
+		require.NoError(t, err)
+
+		s := fakeTLSConn{}
+		err = c.AuthorizeServerConn("dc1", s)
+		require.NoError(t, err)
+	})
+
 }
 
 type fakeTLSConn struct {

--- a/tlsutil/generate.go
+++ b/tlsutil/generate.go
@@ -143,17 +143,17 @@ func GenerateCA(opts CAOpts) (string, string, error) {
 func GenerateCert(opts CertOpts) (string, string, error) {
 	parent, err := parseCert(opts.CA)
 	if err != nil {
-		return "", "", err
+		return "", "", fmt.Errorf("failed to parse CA: %w", err)
 	}
 
 	signee, pk, err := GeneratePrivateKey()
 	if err != nil {
-		return "", "", err
+		return "", "", fmt.Errorf("failed to generate private key: %w", err)
 	}
 
 	id, err := keyID(signee.Public())
 	if err != nil {
-		return "", "", err
+		return "", "", fmt.Errorf("failed to get keyID from public key: %w", err)
 	}
 
 	sn := opts.Serial
@@ -185,7 +185,7 @@ func GenerateCert(opts CertOpts) (string, string, error) {
 
 	bs, err := x509.CreateCertificate(rand.Reader, &template, parent, signee.Public(), opts.Signer)
 	if err != nil {
-		return "", "", err
+		return "", "", fmt.Errorf("failed to create certificate: %w", err)
 	}
 
 	var buf bytes.Buffer

--- a/website/content/docs/agent/options.mdx
+++ b/website/content/docs/agent/options.mdx
@@ -2276,11 +2276,14 @@ signed by the CA can be used to gain full access to Consul.
   endpoint was created by the CA that the consul client was setup with. If the
   UI is served, the same checks are performed.
 
-- `verify_incoming_rpc` - If set to true, Consul
-  requires that all incoming RPC connections make use of TLS and that the client
+- `verify_incoming_rpc` - When set to true, Consul
+  requires that all incoming RPC connections use TLS and that the client
   provides a certificate signed by a Certificate Authority from the [`ca_file`](#ca_file)
   or [`ca_path`](#ca_path). By default, this is false, and Consul will not enforce
   the use of TLS or verify a client's authenticity.
+
+  ~> **Security Note:** `verify_incoming_rpc` _must_ be set to true to prevent anyone
+  with access to the RPC port from gaining full access to the Consul cluster.
 
 - `verify_incoming_https` - If set to true,
   Consul requires that all incoming HTTPS connections make use of TLS and that the
@@ -2303,14 +2306,12 @@ signed by the CA can be used to gain full access to Consul.
   server unencrypted is to also enable `verify_incoming` which requires client
   certificates too.
 
-- `verify_server_hostname` - If set to true,
-  Consul verifies for all outgoing TLS connections that the TLS certificate
-  presented by the servers matches `server.<datacenter>.<domain>`
-  hostname. By default, this is false, and Consul does not verify the hostname
-  of the certificate, only that it is signed by a trusted CA. This setting is
-  _critical_ to prevent a compromised client from being restarted as a server
-  and having all cluster state _including all ACL tokens and Connect CA root keys_
-  replicated to it. This is new in 0.5.1.
+- `verify_server_hostname` - When set to true, Consul verifies the TLS certificate
+  presented by the servers match the hostname `server.<datacenter>.<domain>`.
+  By default this is false, and Consul does not verify the hostname
+  of the certificate, only that it is signed by a trusted CA. This setting _must_ be enabled
+  to prevent a compromised client from gaining full read and write access to all
+  cluster data _including all ACL tokens and Connect CA root keys_. This is new in 0.5.1.
 
   ~> **Security Note:** From versions 0.5.1 to 1.4.0, due to a bug, setting
   this flag alone _does not_ imply `verify_outgoing` and leaves client to server

